cpp/src/text/minhash.cu

```diff
@@ -0,0 +1,188 @@
+
+    // initialize hashes output for this string
+    if (lane_idx == 0) {
+      auto const init = d_str.empty() ? 0 : std::numeric_limits<cudf::hash_value_type>::max();
+      thrust::fill(thrust::seq, d_output, d_output + seeds.size(), init);
+    }
+    __syncwarp();
+
+    auto const begin = d_str.begin() + lane_idx;
+    auto const end   = [d_str, width = width] {
+      auto const length = d_str.length();
+      if (length > width) { return (d_str.end() - (width - 1)); }
+      return d_str.begin() + static_cast<cudf::size_type>(length > 0);
+    }();
+
+    // each lane hashes substrings of the given width
+    for (auto itr = begin; itr < end; itr += cudf::detail::warp_size) {
+      auto const offset = itr.byte_offset();
+      auto const hash_str =
+        cudf::string_view(d_str.data() + offset, (itr + width).byte_offset() - offset);
+
+      // hashing each seed on the same section of string is 10x faster than
+      // computing the substrings for each seed
+      for (std::size_t seed_idx = 0; seed_idx < seeds.size(); ++seed_idx) {
+        auto const hasher = cudf::detail::MurmurHash3_32<cudf::string_view>{seeds[seed_idx]};
+        auto const hvalue = hasher(hash_str);
+        atomicMin(d_output + seed_idx, hvalue);
+      }
+    }
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<cudf::column> minhash(cudf::strings_column_view const& input,
+                                      cudf::device_span<cudf::hash_value_type const> seeds,
+                                      cudf::size_type width,
+                                      cudf::hash_id hash_function,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(!seeds.empty(), "Parameter seeds cannot be empty", std::invalid_argument);
+  CUDF_EXPECTS(width >= 2,
+               "Parameter width should be an integer value of 2 or greater",
+               std::invalid_argument);
+  CUDF_EXPECTS(hash_function == cudf::hash_id::HASH_MURMUR3,
+               "Only murmur3 hash algorithm supported",
+               std::invalid_argument);
+
+  auto output_type = cudf::data_type{cudf::type_to_id<cudf::hash_value_type>()};
+  if (input.is_empty()) { return cudf::make_empty_column(output_type); }
+
+  auto const d_strings = cudf::column_device_view::create(input.parent(), stream);
+
+  auto hashes   = cudf::make_numeric_column(output_type,
+                                          input.size() * static_cast<cudf::size_type>(seeds.size()),
+                                          cudf::mask_state::UNALLOCATED,
+                                          stream,
+                                          mr);
+  auto d_hashes = hashes->mutable_view().data<cudf::hash_value_type>();
+
+  thrust::for_each_n(rmm::exec_policy(stream),
+                     thrust::counting_iterator<cudf::size_type>(0),
+                     input.size() * cudf::detail::warp_size,
```

Comment:
Similar query about signed overflow here.

Reply:
This one is different. This could definitely happen for a large number of smallish strings which is not expected for this API but certainly possible. Changed this to use `std::size_t` for the indices instead.